docs/design_proposals/filesystem-resolution-proposal-01.md

```diff
@@ -0,0 +1,110 @@
+creates a tar file which contains the entire content of a given layer in the
+overlay filesystem. Each overlay layer corresponds to one image layer.
+
+Overlay filesystems should only contain the objects changed in each layer;
+meaning that if only one file changes between some layer A and some B, layer B
+would only contain a single file (the one that changed).
+
+To accomplish this, Kaniko walks the entire filesystem to discover every object.
+Some of these objects may actually be a symlink to another object in the
+filesystem; in these cases we must consider both the link and the target object.
+
+Kaniko also maintains a set of whitelisted (aka ignored) filepaths. Any object
+which matches one of these filepaths should be ignored by kaniko.
+
+This results in a 3 dimensional search space
+
+* changed relative to previous layer
+* symlink
+* whitelisted
+
+Kaniko must also track which objects are referred to by multiple stages; this
+functionality is out of scope for this proposal.
+
+This search space is currently managed in an inconsistent and somewhat ad-hoc
+way; code that manages the various search dimensions is spread out and
+duplicated. There are also a number of poorly handled edge cases which continue
+to cause bugs.
+
+The search space dimensions cannot be reduced or substituted.
+
+Currently there are a number of bugs around symlinks incorrectly resolved,
+whitelists not respected, and unchanged files added to layers.
+
+## Design
+
+Resolution of the filesystem and the objects it contains should be handled
+consistently and with a single API.
+
+* Callers of this API should not be concerned with the type of object at a given filepath (e.g. symlink or not).
+* Callers of this API should not be concerned with whether a given path is whitelisted.
+* This API should return a set of filepaths which should be added to the layer
+  without further processing.
+
+The API should take a limited set of arguments
+* The root path in the filesystem
+* The whitelist
+
+The API should return only two arguments
+* A set of filepaths which should be added to the layer without further
+  processing
+* error
+
+### Open Issues/Questions
+
+\<Ignore symlinks targeting whitelisted paths?\>
+
+Resolution: Not Yet Resolved
+
+## Implementation plan
+
+* Write the new API
+* Write tests for the new API
+* Integrate the new API into existing code
+
+## Integration test plan
+
+Add integration tests to the existing suite which cover the known bugs
+
+## Notes
+
+Given some path `/usr/lib/foo` which is a link to `/etc/foo/`
+
+And `/etc/foo` contains `/etc/foo/bar.txt`
+
+Adding a link `/usr/lib/foo/bar.txt` => `/etc/foo/bar.txt` will break the image
+
+In a linux shell this raises an error
+```
+$ ls /usr/lib/bar
+=> /usr/lib/bar/foo.txt
+$ ln -s /usr/lib/bar barlink
+$ ln -s /usr/lib/bar/foo.txt barlink/foo.txt
+=> ERROR
+```
+
+Given some path `/usr/foo/bar` which is a link to `/dev/null`, and `/dev` is
```

Comment:
What happens when a docker files has a command
`Copy some_bin /var/run/some_bin`

`/var/run` is whitelisted.
In this scenario, `Docker` would create `/var/run/some_bin` in the layer.

Reply:
I'm not sure, what do we currently do when the destination of a copy command is a whitelisted path?

Reply:
Right now, for Copy and ADD command we do the following
- Resolve source and dest by substituting any env varaibles
- Resolve any wild cards in sources
- Get Relative source file path respect to root path. In this step, 
   - we apply [`CheckWhitelist(path) && !HasFilepathPrefix(path, root, false)`](https://github.com/GoogleContainerTools/kaniko/blob/f3b2c4064be52706f892799d762c446541dcc378/pkg/util/fs_util.go#L439)
   - if `/var/run` is ignored, then this condition should return true for `/var/run/some_bin`
   - if `/var/run` is ignores then this condition should return false for `var/run/another_dir/some_bin`

We should simplify this.

For symlink dir in whitelisted path, there is another flow.

Reply:
So what is the end result? Do we copy the file to the whitelist destination?

Reply:
I did a quick test
```
FROM ubuntu:16.04
COPY meow.txt /var/run
```
docker
```
ls /var/run
=> initctl  lock  log  meow.txt  mount  sendsigs.omit.d  shm  systemd  user  utmp
```
kaniko
```
ls /var/run
/var/run
```

Reply:
Interestingly

I did a quick test
```
FROM ubuntu:16.04
COPY meow.txt /var/run/meow.txt
```
docker
```
ls /var/run
=> initctl  lock  log  meow.txt  mount  sendsigs.omit.d  shm  systemd  user  utmp
```
kaniko
```
ls /var/run
=> /var/run/meow.txt
```